x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/kerberos/support/KerberosTicketValidator.java

```diff
@@ -0,0 +1,271 @@
+    }
+
+    /**
+     * Privileged Wrapper that invokes action with Subject.doAs to perform work as
+     * given subject.
+     *
+     * @param subject {@link Subject} to be used for this work
+     * @param action {@link PrivilegedExceptionAction} action for performing inside
+     *            Subject.doAs
+     * @return the value returned by the PrivilegedExceptionAction's run method
+     * @throws PrivilegedActionException
+     */
+    private static <T> T doAsWrapper(final Subject subject, final PrivilegedExceptionAction<T> action) throws PrivilegedActionException {
+        try {
+            return AccessController.doPrivileged((PrivilegedExceptionAction<T>) () -> Subject.doAs(subject, action));
+        } catch (PrivilegedActionException pae) {
+            if (pae.getCause() instanceof PrivilegedActionException) {
+                throw (PrivilegedActionException) pae.getCause();
+            }
+            throw pae;
+        }
+    }
+
+    /**
+     * Privileged wrapper for closing GSSContext, does not throw exceptions but logs
+     * them as a debug message.
+     *
+     * @param gssContext GSSContext to be disposed.
+     */
+    private static void privilegedDisposeNoThrow(final GSSContext gssContext) {
+        if (gssContext != null) {
+            try {
+                AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
+                    gssContext.dispose();
+                    return null;
+                });
+            } catch (PrivilegedActionException e) {
+                LOGGER.debug("Could not dispose GSS Context", (Exception) ExceptionsHelper.unwrapCause(e));
+            }
+        }
+    }
+
+    /**
+     * Privileged wrapper for closing LoginContext, does not throw exceptions but
+     * logs them as a debug message.
+     *
+     * @param loginContext LoginContext to be closed
+     */
+    private static void privilegedLogoutNoThrow(final LoginContext loginContext) {
+        if (loginContext != null) {
+            try {
+                AccessController.doPrivileged((PrivilegedExceptionAction<Void>) () -> {
+                    loginContext.logout();
+                    return null;
+                });
+            } catch (PrivilegedActionException e) {
+                LOGGER.debug("Could not close LoginContext", (Exception) ExceptionsHelper.unwrapCause(e));
+            }
+        }
+    }
+
+    /**
+     * Performs authentication using provided keytab
+     *
+     * @param keytabFilePath Keytab file path
+     * @param krbDebug if {@code true} enables jaas krb5 login module debug logs.
+     * @return authenticated {@link LoginContext} instance. Note: This needs to be
+     *         closed using {@link LoginContext#logout()} after usage.
+     * @throws PrivilegedActionException when privileged action threw exception
+     */
+    private static LoginContext serviceLogin(final String keytabFilePath, final boolean krbDebug)
```

Comment:
Nit: You're inconsistent in whether it's a `keytab` or a `keyTab`. I have no opinion, but pick one :)